docs: additional differential loading clarification (#34620)
angular/angular: aio/content/guide/deployment.md

Even as JavaScript continues to evolve, with new features being introduced, not all browsers are updated with support for these new features at the same pace.

The code you write in development using TypeScript is compiled and bundled into ES2015, the JavaScript syntax that is compatible with most browsers.

> All modern browsers support ES2015 and beyond, but in most cases, you still have to account for users accessing your application from a browser that does not support ES2015.

Reviewer: I'd split this into two smaller sentences:
```suggestion
All modern browsers support ES2015 and beyond. For some applications, you still have to account for users accessing your application from a browser that does not support ES2015.
```

> When targeting browsers that do not support ES2015, like IE11, you will need to do two things to ensure that the application runs correctly:

Reviewer 2: like -> such as
you will need to -> you must

>   1. Downlevel all JavaScript to ES5 so that the code is syntactically compatible

Reviewer 2: add period -- compatible.

  2. Load [polyfills](guide/browser-support#polyfills) to fill in any missing modern browser APIs that your application needs.

> Differential loading is a technique of building and loading your application that meets both of these requirements, without penalizing users with modern browsers by having them download code that only older browsers need. When differential loading is enabled (which is the default) the CLI builds two separate bundles as part of your deployed application.

Reviewer 2: *Differential loading*
Introduction of term in italics -- UNLESS you add it to the glossary and make it a link (which is not a bad idea).
I can help do that if you want -- in that case, `[Differential loading](guide/glossary#differential_loading)`

Author: Text updated here to reference the glossary.
Glossary update PR: https://github.com/angular/angular/pull/34884

> [Differential loading](guide/glossary#differential-loading) is a technique of building and loading your application that meets both of these requirements, without penalizing users with modern browsers by having them download code that only older browsers need. When differential loading is enabled (which is the default) the CLI builds two separate bundles as part of your deployed application.

Reviewer 3: I had to read to all the way down here to learn that:

1. the differential serving is something the CLI does and solves this compatibility problem with older browsers
2. this feature is enabled in the CLI by default, so I don't need to do anything to get it to work

This information is crucial and should be provided to the user upfront. I suggest that you rewrite this section from the customer's perspective. The customer doesn't care about how this stuff is implemented. What they care about is the following:

- Angular CLI by default ensures that your application is compatible with legacy browsers, such as IE11
- to do this, Angular CLI implements a feature called Differential Loading
- this feature requires extra processing and slows down the build, so if the audience of the app doesn't use IE11, then this feature can be disabled by doing X, to speed up the build
- differential loading is disabled in dev mode (when `ng test` or `ng serve`) are run, so if the developer needs to test their app on IE11 or other old browsers, then they can enable differential loading in the dev mode by doing Y
- *lastly* add a section on what differential serving is and how it works for those that want to know about the implementation details.

Reviewer: I will defer to Igor on this, but I do agree: writing with the user in mind is much better than writing for what the feature/product does. I am happy to help with this if you need someone to plan on outline or provide more structural edits.

Author: My initial goal was to adjust some of the language to be more technically correct and to provide some additional insight into the underlying functionality.  This was driven by issue reports from users that were unsure of the intended/expected behavior when enabled.

However, I do agree that the section as a whole could be better structured.  I think if a larger restructuring where to occur, it may also be better served as either a standalone guide or as part of the browser support guide.  Browser support is the main goal of differential loading and there is already a short section in that guide discussing it.
Thoughts?

The code you write in development using TypeScript is compiled and bundled into ES2015, the JavaScript syntax that is compatible with most browsers.
All modern browsers support ES2015 and beyond, but in most cases, you still have to account for users accessing your application from a browser that doesn't.

> When targeting older browsers, [polyfills](guide/browser-support#polyfills) can help bridge the gap by providing functionality that doesn't exist in the older versions of JavaScript supported by those browsers. To fully bridge the gap, you can compile a version of the application directly to ES5, in addition to the primary version compiled to ES2015.

Author: @jbogarthyde I did an initial update based on the proposal above.  Can you do another review of this section?

Users with modern browsers, however, shouldn't have to pay the price of increased bundle size that comes with both the polyfills they don't need and the larger code size for ES5 equivalents of ES2015 code. For example, ES5 does not support classes and they must be transformed into ES5 code that mimics their behavior.
Differential loading, which is supported by default in Angular CLI version 8 and higher, solves this problem.

Differential loading is a strategy that allows your web application to support multiple browsers, but only load the necessary code that the browser needs. When differential loading is enabled (which is the default) the CLI builds two separate bundles as part of your deployed application.

* The first bundle contains modern ES2015 syntax, takes advantage of built-in support in modern browsers, ships fewer polyfills, and results in a smaller bundle size.

* The second bundle contains code in the old ES5 syntax, along with all necessary polyfills for Angular to function. This results in a larger bundle size, but supports older browsers.

> To optimize differential loading build performance, the CLI uses a bundle level transformation process.

Reviewer 2: bundle-level

> This process requires that only one full compilation of the application in ES2015 syntax to take place.

Reviewer 2: requires only one full compilation of the application in ES2015 syntax.

Once this full ES2015 compilation is complete, a post-processing step is performed that transforms each output JavaScript file into an additional ES5 variant. An additional ES5 only polyfills file is also generated that contains the required polyfills for ES5 browsers.

> This process also guarantees that the entire bundled application is ES5 compliant including third-party code.

Reviewer 2: 490:
 Once the full ES2015 compilation is complete, a post-processing step transforms each output JavaScript file into an additional ES5 variant, and generates an additional ES5-only file containing the polyfills that are required for ES5 browsers.
491: (new paragraph)
 The bundle-level transformation process guarantees that the entire bundled application is ES5 compliant, including any third-party code. This is not the case when you simply change the compilation target to `es5`.

> This process is different than changing the compilation target to `es5`.

Reviewer 2: Bundle-level transformation is more efficient and thorough than simply changing the compilation target to `es5`.

>  The bundle-level transformation process guarantees that the entire bundled application is ES5 compliant, including any third-party code. This is not the case when you simply change the compilation target to `es5`.

Reviewer 3: ```suggestion
 The bundle-level transformation process guarantees that the entire bundled application is ES5 compliant, including any third-party code. This is not the case when you simply change the compilation target to `es5` in `tsconfig.app.json`.
```

> Bundle-level transformation is more efficient and thorough than simply changing the compilation target to `es5`.

Reviewer 3: I don't think that this line and all the other newly added lines below are adding much value and only confuse the reader. I suggest that we remove them.

Author: Had similar thoughts when writing this part.  My initial reasoning for including it was that there have been several issues expressing confusion on why user's applications worked with differential loading but failed when only targeting es5 in the tsconfig.  (Upon investigation, some of the third-party libraries didn't provide ES5 code).

However, the underlying issue is something that will be corrected soon (and prevent the aforementioned situation) so I think removing it as you suggested is a good idea.

> When changing the compilation target to `es5`, only the TypeScript application code will be transformed to ES5. For libraries, the ES5 entry point of the library's package will be used directly. If a library does not have one then the main entry point of the library's package will be used instead.

Reviewer 2: When you change the compilation target to `es5`, only the TypeScript application code is transformed to ES5. For libraries, the ES5 entry point of the library's package is used directly if one exists. If it does not, the main entry point of the library's package is used, and that entry point might contain code that is not ES5-compatible.

> The main entry point may not contain code that is ES5 compatible.

Reviewer 2: remove (added to previous sentence)

This can result in a bundled application that is a mix of ES5 and ES2015+ code.

> If using a compilation target of `es5`, it is important to ensure that all libraries used by the application provide ES5 compatible code.

Reviewer 2: If you do use an `es5` compilation target (rather that the bundle-level transformation), you must ensure that all libraries used by the application provide ES5 compatible code.

> If you do use an `es5` compilation target (rather that the bundle-level transformation), you must ensure that all libraries used by the application provide ES5 compatible code.

Reviewer 2: rather than

## Local development in older browsers

In Angular CLI version 8 and higher, differential loading is enabled by default for the `ng build` command.

> The `ng serve`, `ng test`, and `ng e2e` commands, however, generate a single ES2015 build that cannot run in older browsers, such as IE 11.

Reviewer: ```suggestion
By default, the `ng serve`, `ng test`, and `ng e2e` commands, generate a single ES2015 build that cannot run in older browsers, such as IE 11.
```

> The `ng serve`, `ng test`, and `ng e2e` commands, however, generate a single ES2015 build which cannot run in older browsers, such as IE 11.

Reviewer 2: which --> that
